### PR TITLE
feat: representation and overview for group nodes and edges

### PR DESCRIPTION
**Pull request by LauraBoenchenLB**
Adds 
- Summary for attributes based on their type
- overview of nodes contained in groups of a medrecord and their attributes
- overview of edges contained in groups or edges connecting group nodes and their attributes
- representation of the medrecord based on these nodes and edges

### CODE REVIEW

File: medmodels/medrecord/medrecord.py [hunk 1]
```diff
@@ -1301,3 +1302,126 @@ def __getitem__(
...
+                schema={
+                    "Edges": pl.String,
+                    "Count": pl.String,
+                    "Attribute": pl.String,
+                    "Info": pl.String,
+                }
+            )
+        ]
+
+        for group1 in self.groups:
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/medrecord.py:1372]: better name for `group1`?

File: medmodels/medrecord/medrecord.py [hunk 2]
```diff
@@ -1301,3 +1302,126 @@ def __getitem__(
...
+                    [
+                        pl.lit(group1).alias("Edges"),
+                        pl.lit(str(len(edges))).alias("Count"),
+                        pl.all(),
+                    ]
+                )
+                edge_groups.append(edge_info)
+
+            # edges connecting different groups
+            for group2 in self.groups:
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/medrecord.py:1394]: better name for `group2`?

File: medmodels/medrecord/medrecord.py [hunk 3]
```diff
@@ -1301,3 +1302,126 @@ def __getitem__(
...
+                            pl.lit(f"{group1} -> {group2}").alias("Edges"),
+                            pl.lit(str(len(edges))).alias("Count"),
+                            pl.all(),
+                        ]
+                    )
+                    edge_groups.append(edge_info)
+        edge_table = pl.concat(edge_groups)
+        if edge_table.is_empty():
+            edge_table = pl.DataFrame({col: "-" for col in edge_table.columns})
+        return edge_table
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/medrecord.py:1412]:
```suggestion
        for group1 in self.groups:
            # edges in groups
            edges = self.group(group1)["edges"]
            
            if edges:
                schema = (
                    self.schema.group(group1).edges
                    if group1 in self.schema.groups
                    else None
                )

                edge_info = extract_attribute_summary(self.edge[edges], schema)

                edge_info = edge_info.select(
                    [
                        pl.lit(group1).alias("Edges"),
                        pl.lit(str(len(edges))).alias("Count"),
                        pl.all(),
                    ]
                )

                edge_groups.append(edge_info)

            # edges connecting different groups
            for group2 in self.groups:
                edges = self.select_edges(
                    edge().connected_source_with(node().in_group(group1))
                    & edge().connected_target_with(node().in_group(group2))
                )

                if not edges:
                    continue

                edge_info = extract_attribute_summary(self.edge[edges])
                edge_info = edge_info.select(
                    [
                        pl.lit(f"{group1} -> {group2}").alias("Edges"),
                        pl.lit(str(len(edges))).alias("Count"),
                        pl.all(),
                    ]
                )

                edge_groups.append(edge_info)

        edge_table = pl.concat(edge_groups)

        if edge_table.is_empty():
            edge_table = pl.DataFrame({col: "-" for col in edge_table.columns})

        return edge_table
```

File: medmodels/medrecord/medrecord.py [hunk 4]
```diff
@@ -1301,3 +1302,126 @@ def __getitem__(
...
+    def __repr__(self) -> str:
+        representation = prettify_table(self.describe_group_nodes())
+        representation.append("")
+        representation.extend(prettify_table(self.describe_group_edges()))
+
+        return "\n".join(representation)
+
+    def show_group_nodes(self):
+        nodes_table = prettify_table(self.describe_group_nodes())
+        [print(line) for line in nodes_table]
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/medrecord.py:1423]: one print with `\n` join would be cleaner, same as in `__repr__`

File: medmodels/medrecord/medrecord.py [hunk 5]
```diff
@@ -1301,3 +1302,126 @@ def __getitem__(
...
+
+    def show_group_edges(self):
+        edges_table = prettify_table(self.describe_group_edges())
+        [print(line) for line in edges_table]
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/medrecord.py:1427]: one print with `\n` join would be cleaner, same as in `__repr__`

File: medmodels/medrecord/medrecord.py [hunk 6]
```diff
@@ -1301,3 +1302,126 @@ def __getitem__(
...
+        return edge_table
+
+    def __repr__(self) -> str:
+        representation = prettify_table(self.describe_group_nodes())
+        representation.append("")
+        representation.extend(prettify_table(self.describe_group_edges()))
+
+        return "\n".join(representation)
+
+    def show_group_nodes(self):
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/medrecord.py:1421]: maybe `print_group_nodes`? Altough the `group_nodes` part would still sound a bit weird to me. I'd not know what to expect from this function.

Also missing docstrings

File: medmodels/medrecord/medrecord.py [hunk 7]
```diff
@@ -1301,3 +1302,126 @@ def __getitem__(
...
+        nodes_table = prettify_table(self.describe_group_nodes())
+        [print(line) for line in nodes_table]
+
+    def show_group_edges(self):
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/medrecord.py:1425]: See `show_group_nodes` comment

File: medmodels/medrecord/medrecord.py [hunk 8]
```diff
@@ -1301,3 +1302,126 @@ def __getitem__(
...
+                        pl.lit(group).alias("Nodes"),
+                        pl.lit(str(len(nodes))).alias("Count"),
+                        pl.all(),
+                    ]
+                )
+                node_groups.append(node_info)
+        node_table = pl.concat(node_groups)
+        if node_table.is_empty():
+            node_table = pl.DataFrame({col: "-" for col in node_table.columns})
+        return node_table
```
[the remark this answers is not in the document]
Comment (JabobKrauskopf) [reply medmodels/medrecord/medrecord.py:1349]: added early returns and a bit of cleanup

File: medmodels/medrecord/medrecord.py (hunk 3 above)
Comment (JabobKrauskopf) [reply medmodels/medrecord/medrecord.py:1412]: added early returns and a bit of cleanup

File: medmodels/medrecord/overview.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            ## Without Schema
+            else:
+                if data[attribute].dtype.is_numeric():
+                    attribute_info = [
+                        f"min: {data[attribute].min()}",
+                        f"max: {data[attribute].max()}",
+                        f"mean: {data[attribute].mean():.{decimal}f}",
+                    ]
+                else:
+                    try:
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/overview.py:77]: `try ... except` is bad practice when other options are available. Try using `data[attribute].str.to_datetime(strict=False)` for example which would not raise an error

File: medmodels/medrecord/overview.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                            attribute_info = [f"{len(categories)} unique values"]
+                        else:
+                            attribute_info = [f"Values: {', '.join(list(categories))}"]
+
+            data_dict["Attribute"].extend([attribute] * len(attribute_info))
+            data_dict["Info"].extend(attribute_info)
+    else:
+        [data_dict[x].append("-") for x in data_dict.keys()]
+
+    return pl.DataFrame(data_dict)
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/overview.py:96]:
```suggestion
    attributes.sort()

    if not attributes:
        return pl.DataFrame({"Attribute": ["-"], "Info": ["-"]})

    for attribute in attributes:
        if schema and attribute in schema:
            if schema[attribute][1] == AttributeType.Continuous:
                attribute_info = [
                    f"min: {data[attribute].min()}",
                    f"max: {data[attribute].max()}",
                    f"mean: {data[attribute].mean():.{decimal}f}",
                ]

            elif schema[attribute][1] == AttributeType.Temporal:
                attribute_info = [
                    f"min: {min(data[attribute])}",
                    f"max: {max(data[attribute])}",
                ]

            elif schema[attribute][1] == AttributeType.Categorical:
                categories = data[attribute].unique().sort()
                if len(categories) > 5:
                    attribute_info = [f"{len(categories)} categories"]
                else:
                    category_str = ", ".join(categories)
                    if len(category_str) < 100:
                        attribute_info = [f"Categories: {', '.join(list(categories))}"]
                    else:
                        attribute_info = [f"{len(categories)} categories"]
            else:
                attribute_info = [f"{len(data[attribute].unique())} unique values"]

        ## Without Schema
        else:
            if data[attribute].dtype.is_numeric():
                attribute_info = [
                    f"min: {data[attribute].min()}",
                    f"max: {data[attribute].max()}",
                    f"mean: {data[attribute].mean():.{decimal}f}",
                ]
            else:
                try:
                    time_attribute = data[attribute].str.to_datetime()
                    attribute_info = [
                        f"min: {min(time_attribute)}",
                        f"max: {max(time_attribute)}",
                    ]
                except pl.exceptions.ComputeError:
                    categories = data[attribute].unique().sort()
                    category_str = ", ".join(categories)
                    if (len(categories) > 5) | (len(category_str) > 100):
                        attribute_info = [f"{len(categories)} unique values"]
                    else:
                        attribute_info = [f"Values: {', '.join(list(categories))}"]

        data_dict["Attribute"].extend([attribute] * len(attribute_info))
        data_dict["Info"].extend(attribute_info)

    return pl.DataFrame(data_dict)
```

Comment (JabobKrauskopf) [reply medmodels/medrecord/overview.py:96]: also added early return

File: medmodels/medrecord/overview.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def prettify_table(table_info: pl.DataFrame) -> List[str]:
+    """Takes a DataFrame and turns it into a list for printing a pretty table.
+
+    Args:
+        table_info (pl.DataFrame): Table in DataFrame format.
+
+    Returns:
+        List[str]: _description_
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/overview.py:106]: `_description_`

File: medmodels/medrecord/tests/test_overview.py
Comment (JabobKrauskopf) [inline medmodels/medrecord/tests/test_overview.py:1]: also test the other functions by comparing the strings or by mocking `sys.stdout`

File: medmodels/medrecord/overview.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+                        category_str = ", ".join(categories)
+                        if (len(categories) > 5) | (len(category_str) > 100):
+                            attribute_info = [f"{len(categories)} unique values"]
+                        else:
+                            attribute_info = [f"Values: {', '.join(list(categories))}"]
+
+            data_dict["Attribute"].extend([attribute] * len(attribute_info))
+            data_dict["Info"].extend(attribute_info)
+    else:
+        [data_dict[x].append("-") for x in data_dict.keys()]
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/overview.py:94]: list comprehensions should not be used this way

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]
Very nice output! Me likey.

Some adjustments for now

File: medmodels/medrecord/overview.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+    return pl.DataFrame(data_dict)
+
+
+def prettify_table(table_info: pl.DataFrame) -> List[str]:
```
Comment (MarIniOnz) [inline medmodels/medrecord/overview.py:99]: I do not quite like this naming, maybe better _embelish_table_

Comment (JabobKrauskopf) [reply medmodels/medrecord/overview.py:99]: _prettify_ is a common term for similar things to this in the programming context :)

File: medmodels/medrecord/overview.py (hunk 9 above)
Comment (LauraBoenchenLB) [reply medmodels/medrecord/overview.py:77]: data[attribute].str.to_datetime(strict=False) does not work in the case of not having any datetime object in the series.

File: medmodels/medrecord/medrecord.py [hunk 14]
```diff
@@ -1301,3 +1302,176 @@ def __getitem__(
...
+        return edge_table
+
+    def __repr__(self) -> str:
+        representation = prettify_table(self._describe_group_nodes())
+        representation.append("")
+        representation.extend(prettify_table(self._describe_group_edges()))
+
+        return "\n".join(representation)
+
+    def print_attribute_table_nodes(self) -> None:
```
Comment (MarIniOnz) [inline medmodels/medrecord/medrecord.py:1435]: I dont quite like the name of the function. Maybe?
```suggestion
    def print_table_nodes(self) -> None:
```

File: medmodels/medrecord/medrecord.py [hunk 15]
```diff
@@ -1301,3 +1302,176 @@ def __getitem__(
...
+                        gender      Categories: F, M
+        ----------------------------------------------
+
+        """
+
+        nodes_table = prettify_table(self._describe_group_nodes())
+
+        print("\n".join(nodes_table))
+
+    def print_attribute_table_edges(self) -> None:
```
Comment (MarIniOnz) [inline medmodels/medrecord/medrecord.py:1457]: Same as above about the naming.

**Review by MarIniOnz — CHANGES_REQUESTED** [review]
Beautiful PR and amazingly looking table! Props!

Minor changes requested and genuine questions I do not know whether they should be addressed or are already addressed. Feel free to answer them and resolve them if you think the issues they address should not be taken into account.

File: medmodels/medrecord/medrecord.py (hunk 14 above)
Comment (LauraBoenchenLB) [reply medmodels/medrecord/medrecord.py:1435]: that's what I had before and Jakob didn't like it :D

File: medmodels/medrecord/medrecord.py [hunk 16]
```diff
@@ -1301,3 +1302,176 @@ def __getitem__(
...
+
+        Returns:
+            pl.DataFrame: Dataframe with all nodes in medrecord groups and their attributes.
+
+        """
+
+        node_groups = [
+            pl.DataFrame(
+                schema={
+                    "Nodes": pl.String,
```
[the remark this answers is not in the document]
Comment (MarIniOnz) [reply medmodels/medrecord/medrecord.py:1315]: Sounds better to me. Node Group and Edges Groups maybe (since they connect two Groups)

**Review by MarIniOnz — CHANGES_REQUESTED** [review]
Minor comments, almost there! Also, there is an unresolved conversation that needs fixing from my previous review.

Additionally, I would suggest we update the version of polars on the dependencies, since on the current version, added tests are failing locally. Any problem with that? @JabobKrauskopf

File: medmodels/medrecord/medrecord.py [hunk 17]
```diff
@@ -1232,3 +1233,168 @@ def __getitem__(
...
+                    "Attribute": pl.String,
+                    "Info": pl.String,
+                }
+            )
+        ]
+
+        groups = sorted(set(self.groups))
+
+        for source_group in groups:
+            # edges in groups
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/medrecord.py:1300]: inconsistent comment (the other function doesnt have that comment). probably not needed

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]
looking good. only some minor stuff. :)
MOSTLY&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SPACING

**Review by MarIniOnz — DISMISSED** [review]

File: medmodels/medrecord/_overview.py [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+def extract_attribute_summary(
+    attribute_dict: Union[Dict[EdgeIndex, Attributes], Dict[NodeIndex, Attributes]],
+    schema: Optional[AttributesSchema] = None,
+    decimal: int = 2,
+) -> pl.DataFrame:
+    """Extracts a summary from a node or edge attribute dictionary.
+
+
+    Example:
+    ┌────────────────┬──────────────────────────┐
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/_overview.py:18]: example is wrong, no?

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]

File: medmodels/medrecord/_overview.py (hunk 18 above)
Comment (LauraBoenchenLB) [reply medmodels/medrecord/_overview.py:18]: why? This is not for the printed table, but the data frame for each individual group.

**Review by JabobKrauskopf — DISMISSED** [review]

**Review by MarIniOnz — APPROVED** [review]

File: medmodels/medrecord/tests/test_overview.py [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+            str_attributes.equals(
+                pl.DataFrame({"Attribute": "ATC", "Info": "Values: B01AA03, B01AF01"})
+            )
+        )
+
+        # nan attribute
+        nan_attributes = extract_attribute_summary(
+            medrecord.node[node().in_group("Aspirin")]
+        )
+        print(nan_attributes)
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/tests/test_overview.py:92]:
```suggestion
```

File: medmodels/medrecord/_overview.py [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+
+    attributes = [col for col in data.columns if col != "id"]
+    attributes.sort()
+
+    if not attributes:
+        return pl.DataFrame({"Attribute": ["-"], "Info": ["-"]})
+
+    for attribute in attributes:
+        attribute_values = data[attribute].drop_nulls()
+        if len(attribute_values) == 0:
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/_overview.py:56]:
```suggestion
        attribute_values = data[attribute].drop_nulls()

        if len(attribute_values) == 0:
```

File: medmodels/medrecord/tests/test_overview.py [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+    ]
+    medrecord = mm.MedRecord.from_pandas(
+        nodes=[(node, "index") for node in nodes],
+        edges=(edges, "source", "target"),
+    )
+    medrecord.add_edges_polars(edges=(edges_disease, "source", "target"))
+    for group, group_list in groups:
+        medrecord.add_group(group, group_list)
+
+    return medrecord
```
Comment (JabobKrauskopf) [inline medmodels/medrecord/tests/test_overview.py:54]: Just FYI: This is les terrible spacing wise

Don't _need_ to fix anything because it's just a test

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]

**Review by MarIniOnz — APPROVED** [review]

**Review by JabobKrauskopf — APPROVED** [review]

**Pull request merged**